### PR TITLE
Significant refactor of mergebot backend.

### PR DESCRIPTION
**Pull request by jasonkuster**
This PR reworks the mergebot backend. Below fixes are largest items, but many QOL fixes throughout.
* Structure reorganized -- mergebot.py is now at top level, with other functionality in mergebot_backend.
* db_publisher added to publish to frontend database.
* Jenkins interactions added to move job verification off of mergebot machine.

This should be reviewed with #13 since they are linked.

### CODE REVIEW

File: config/beam.yaml
Comment (charlesccychen) [inline config/beam.yaml]: Can you fix this?--you indicated this was incorrect on our call.

File: mergebot_backend/log_helper.py
Comment (charlesccychen) [inline mergebot_backend/log_helper.py]: Can you make this argument description clearer?

File: mergebot.py
Comment (charlesccychen) [inline mergebot.py]: This comment should read "used to receive the kill signal".

File: mergebot.py
Comment (charlesccychen) [inline mergebot.py]: nit: consistent capitalization

File: mergebot.py
Comment (charlesccychen) [inline mergebot.py]: nit: consistent capitalization

File: mergebot.py
Comment (charlesccychen) [inline mergebot.py:172]: I am a little worried about the way `mergebot_frontend.db` is instantiated--currently, it is created in the `__init__.py` file of mergebot_frontend/, but this may be unsafe because the forked subprocess retains the same handle to the underlying sqlite database as the parent.  See, for example, the documentation here which indicates this may be unsafe: http://docs.sqlalchemy.org/en/latest/core/connections.html#engine-disposal.  Is it possible to make sure this is (re-)instantiated for each (sub-)process?

File: mergebot_backend/db_publisher.py
Comment (charlesccychen) [inline mergebot_backend/db_publisher.py]: Is the timestamp assign intentionally missing here?

File: mergebot_backend/db_publisher.py
Comment (charlesccychen) [inline mergebot_backend/db_publisher.py]: (Where) do we guarantee that this record is present?

File: mergebot.py
Comment (charlesccychen) [inline mergebot.py]: Do we want to do a blanket exception catch here for the subprocess?

**Review by charlesccychen — COMMENTED** [review]
Thanks!  This looks good, modulo a few comments.

File: config/beam.yaml
Comment (jasonkuster) [reply config/beam.yaml]: Yup, fixed.

File: mergebot.py
Comment (jasonkuster) [reply mergebot.py]: Done.

File: mergebot.py
Comment (jasonkuster) [reply mergebot.py]: Done.

File: mergebot.py
Comment (jasonkuster) [reply mergebot.py]: Done.

File: mergebot.py
Comment (jasonkuster) [reply mergebot.py]: Done.

File: mergebot.py
Comment (jasonkuster) [reply mergebot.py:172]: I will look into ways to make this work. One way I considered was to have a DBPublisher class, whose constructor would create a new handle to the database. Does that sound like a good solution here?

File: mergebot_backend/db_publisher.py
Comment (jasonkuster) [reply mergebot_backend/db_publisher.py]: Think this was just an artifact of refactoring -- we didn't used to check if there was already a row for the poller.

File: mergebot_backend/db_publisher.py
Comment (jasonkuster) [reply mergebot_backend/db_publisher.py]: I see your point -- in the code it's impossible for an item heartbeat to be published without the job_wait status being published first, but a problem with the database could cause that relationship to be violated and we could crash here. I've added some handling for this -- this method can now raise KeyError and its parent handles that appropriately.

File: mergebot_backend/log_helper.py
Comment (jasonkuster) [reply mergebot_backend/log_helper.py]: Done.

File: mergebot.py
Comment (charlesccychen) [reply mergebot.py:172]: Yes, that sounds good.

**Review by charlesccychen — COMMENTED** [review]
Thanks!  This looks good modulo the DB constructor issues.

**Comment by jasonkuster** [comment]
Rebased successfully onto master after frontend merged. Latest commit adds testing version of thread/process-local database connections. Will be testing tomorrow and will ping when ready for final look.

**Comment by jasonkuster** [comment]
@charlesccychen I've verified that the latest commit (e2e00f6) works, and based on the documentation it should solve the threading problems. PTAL.

File: mergebot.py
Comment (charlesccychen) [inline mergebot.py]: Can you move the poller poll into the try block?

File: mergebot_backend/db_publisher.py
Comment (charlesccychen) [inline mergebot_backend/db_publisher.py]: Please add a comment for why this is necessary here.

**Review by charlesccychen — COMMENTED** [review]
Thanks!  LGTM modulo comments.

**Review by charlesccychen — APPROVED** [review]
Thanks!

**Pull request merged**
